File: sniper/relay_buyer.py
"""Buy tokens on any supported chain from a single funding wallet, via
Relay (https://relay.link).

Instead of holding native gas coins on every chain, you fund ONE wallet
on ONE origin chain (e.g. ETH on Base). For each detected token, Relay
quotes a cross-chain swap: we sign and send the origin-chain
transaction(s) it returns, and Relay's solvers deliver the destination
token to the recipient — including SPL tokens on Solana.

Flow: POST /quote -> execute returned origin-chain tx steps -> poll the
status endpoint until the fill succeeds.
"""
import asyncio
import logging

import aiohttp
from web3 import Web3

import config
from sniper.chains import EVM_CHAINS, SOLANA_KEY

log = logging.getLogger("relay_buyer")

RELAY_API = "https://api.relay.link"
NATIVE_EVM = "0x0000000000000000000000000000000000000000"
# Relay's chain id for Solana mainnet.
RELAY_SOLANA_CHAIN_ID = 792703809

STATUS_POLL_SECONDS = 3
STATUS_POLL_TRIES = 60


def destination_chain_id(chain_key: str) -> int | None:
    if chain_key == SOLANA_KEY:
        return RELAY_SOLANA_CHAIN_ID
    chain = EVM_CHAINS.get(chain_key)
    return chain.chain_id if chain else None


def solana_recipient() -> str:
    """Where SPL tokens should land when buying a Solana token cross-chain."""
    if config.SOLANA_RECIPIENT:
        return config.SOLANA_RECIPIENT
    if config.SOLANA_PRIVATE_KEY:
        import base58
        from solders.keypair import Keypair

        return str(Keypair.from_bytes(base58.b58decode(config.SOLANA_PRIVATE_KEY)).pubkey())
    return ""


def build_quote_body(user: str, dest_chain_key: str, token_address: str, amount_wei: int) -> dict:
    origin = EVM_CHAINS[config.RELAY_ORIGIN_CHAIN]
    dest_id = destination_chain_id(dest_chain_key)
    if dest_id is None:
        raise RuntimeError(f"chain {dest_chain_key!r} is not supported via Relay")

    recipient = user
    if dest_chain_key == SOLANA_KEY:
        recipient = solana_recipient()
        if not recipient:
            raise RuntimeError(
                "Solana destination needs SOLANA_RECIPIENT or SOLANA_PRIVATE_KEY set"
            )

    return {
        "user": user,
        "recipient": recipient,
        "originChainId": origin.chain_id,
        "destinationChainId": dest_id,
        "originCurrency": NATIVE_EVM,
        "destinationCurrency": token_address,
        "amount": str(amount_wei),
        "tradeType": "EXACT_INPUT",
        "slippageTolerance": str(config.SLIPPAGE_BPS),
    }


async def buy(dest_chain_key: str, token_address: str, amount_origin_native: float) -> str:
    """Swap `amount_origin_native` of the origin chain's native coin into
    `token_address` on `dest_chain_key`. Returns the last origin tx hash.
    """
    if not config.EVM_PRIVATE_KEY:
        raise RuntimeError("EVM_PRIVATE_KEY is not set (Relay origin wallet)")

    origin = EVM_CHAINS[config.RELAY_ORIGIN_CHAIN]
    w3 = Web3(Web3.HTTPProvider(origin.rpc, request_kwargs={"timeout": 15}))
    account = w3.eth.account.from_key(config.EVM_PRIVATE_KEY)
    amount_wei = w3.to_wei(amount_origin_native, "ether")

    balance = await asyncio.to_thread(w3.eth.get_balance, account.address)
    if balance < amount_wei:
        raise RuntimeError(
            f"insufficient {origin.native_symbol} on {origin.name}: "
            f"have {w3.from_wei(balance, 'ether')}, need {amount_origin_native} plus gas"
        )

    body = build_quote_body(account.address, dest_chain_key, token_address, amount_wei)

    async with aiohttp.ClientSession(trust_env=True) as session:
        async with session.post(
            f"{RELAY_API}/quote", json=body, timeout=aiohttp.ClientTimeout(total=15)
        ) as resp:
            quote = await resp.json()
            if resp.status != 200:
                raise RuntimeError(f"Relay quote failed ({resp.status}): {quote}")

        tx_hash = ""
        check_endpoints: list[str] = []
        for step in quote.get("steps", []):
            if step.get("kind") != "transaction":
                continue
            for item in step.get("items", []):
                data = item.get("data") or {}
                tx_hash = await asyncio.to_thread(_send_origin_tx, w3, account, origin, data)
                log.info("[Relay] step %r sent on %s: %s", step.get("id"), origin.name, tx_hash)
                check = item.get("check") or {}
                if check.get("endpoint"):
                    check_endpoints.append(check["endpoint"])

        if not tx_hash:
            raise RuntimeError(f"Relay quote returned no executable transaction steps: {quote}")

        for endpoint in check_endpoints:
            await _wait_for_fill(session, endpoint)

    return tx_hash


def _to_int(value) -> int:
    """Relay may return numeric fields as int, decimal string, or hex string."""
    if value is None:
        return 0
    if isinstance(value, str):
        return int(value, 16) if value.startswith("0x") else int(value)
    return int(value)


def _send_origin_tx(w3: Web3, account, origin, data: dict) -> str:
    tx = {
        "from": account.address,
        "to": Web3.to_checksum_address(data["to"]),
        "data": data.get("data", "0x"),
        "value": _to_int(data.get("value")),
        "chainId": origin.chain_id,
        "nonce": w3.eth.get_transaction_count(account.address),
    }
    if origin.eip1559:
        base_fee = w3.eth.get_block("latest").get("baseFeePerGas", w3.eth.gas_price)
        priority = w3.to_wei(2, "gwei")
        tx["maxPriorityFeePerGas"] = priority
        tx["maxFeePerGas"] = base_fee * 2 + priority
    else:
        tx["gasPrice"] = w3.eth.gas_price
    tx["gas"] = int(w3.eth.estimate_gas(tx) * 1.3)

    signed = account.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=180)
    if receipt.status != 1:
        raise RuntimeError(f"Relay origin transaction reverted: {tx_hash.hex()}")
    return tx_hash.hex()


async def _wait_for_fill(session: aiohttp.ClientSession, endpoint: str) -> None:
    """Poll Relay's status endpoint until the cross-chain fill completes."""
    url = endpoint if endpoint.startswith("http") else f"{RELAY_API}{endpoint}"
    for _ in range(STATUS_POLL_TRIES):
        await asyncio.sleep(STATUS_POLL_SECONDS)
        try:
            async with session.get(url, timeout=aiohttp.ClientTimeout(total=10)) as resp:
                status = (await resp.json()).get("status", "")
        except (aiohttp.ClientError, asyncio.TimeoutError):
            continue
        if status == "success":
            log.info("[Relay] fill confirmed")
            return
        if status in ("failure", "refund"):
            raise RuntimeError(f"Relay fill ended with status {status!r}")
    raise RuntimeError("Relay fill not confirmed in time (check relay.link/transactions)")

File: src/relayBuyer.ts
/** Buy tokens on any supported chain from a single funding wallet, via
 * Relay (https://relay.link).
 *
 * Instead of holding native gas coins on every chain, you fund ONE wallet
 * on ONE origin chain (e.g. ETH on Base). For each detected token, Relay
 * quotes a cross-chain swap: we sign and send the origin-chain
 * transaction(s) it returns, and Relay's solvers deliver the destination
 * token to the recipient — including SPL tokens on Solana.
 *
 * Flow: POST /quote -> execute returned origin-chain tx steps -> poll the
 * status endpoint until the fill succeeds.
 */
import { JsonRpcProvider, Wallet, getAddress, parseEther, parseUnits, formatEther } from "ethers";

import { config } from "./config.js";
import { EVM_CHAINS, EvmChain, SOLANA_KEY, rpcFor } from "./chains.js";
import { makeLog } from "./log.js";

const log = makeLog("relay_buyer");

export const RELAY_API = "https://api.relay.link";
export const NATIVE_EVM = "0x0000000000000000000000000000000000000000";
// Relay's chain id for Solana mainnet.
export const RELAY_SOLANA_CHAIN_ID = 792703809;

const STATUS_POLL_SECONDS = 3;
const STATUS_POLL_TRIES = 60;

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export function destinationChainId(chainKey: string): number | null {
  if (chainKey === SOLANA_KEY) return RELAY_SOLANA_CHAIN_ID;
  return EVM_CHAINS[chainKey]?.chainId ?? null;
}

/** Where SPL tokens should land when buying a Solana token cross-chain. */
export async function solanaRecipient(): Promise<string> {
  if (config.SOLANA_RECIPIENT) return config.SOLANA_RECIPIENT;
  if (config.SOLANA_PRIVATE_KEY) {
    const [{ Keypair }, bs58] = await Promise.all([
      import("@solana/web3.js"),
      import("bs58").then((m) => m.default),
    ]);
    return Keypair.fromSecretKey(bs58.decode(config.SOLANA_PRIVATE_KEY)).publicKey.toBase58();
  }
  return "";
}

export async function buildQuoteBody(
  user: string,
  destChainKey: string,
  tokenAddress: string,
  amountWei: bigint
): Promise<Record<string, unknown>> {
  const origin = EVM_CHAINS[config.RELAY_ORIGIN_CHAIN];
  const destId = destinationChainId(destChainKey);
  if (destId === null) {
    throw new Error(`chain '${destChainKey}' is not supported via Relay`);
  }

  let recipient = user;
  if (destChainKey === SOLANA_KEY) {
    recipient = await solanaRecipient();
    if (!recipient) {
      throw new Error("Solana destination needs SOLANA_RECIPIENT or SOLANA_PRIVATE_KEY set");
    }
  }

  return {
    user,
    recipient,
    originChainId: origin.chainId,
    destinationChainId: destId,
    originCurrency: NATIVE_EVM,
    destinationCurrency: tokenAddress,
    amount: amountWei.toString(),
    tradeType: "EXACT_INPUT",
    slippageTolerance: String(config.SLIPPAGE_BPS),
  };
}

/** Swap `amountOriginNative` of the origin chain's native coin into
 * `tokenAddress` on `destChainKey`. Returns the last origin tx hash.
 */
export async function buy(
  destChainKey: string,
  tokenAddress: string,
  amountOriginNative: number
): Promise<string> {
  if (!config.EVM_PRIVATE_KEY) {
    throw new Error("EVM_PRIVATE_KEY is not set (Relay origin wallet)");
  }

  const origin = EVM_CHAINS[config.RELAY_ORIGIN_CHAIN];
  const provider = new JsonRpcProvider(rpcFor(origin), origin.chainId, { staticNetwork: true });
  const wallet = new Wallet(config.EVM_PRIVATE_KEY, provider);
  const amountWei = parseEther(String(amountOriginNative));

  const balance = await provider.getBalance(wallet.address);
  if (balance < amountWei) {
    throw new Error(
      `insufficient ${origin.nativeSymbol} on ${origin.name}: ` +
        `have ${formatEther(balance)}, need ${amountOriginNative} plus gas`
    );
  }

  const body = await buildQuoteBody(wallet.address, destChainKey, tokenAddress, amountWei);

  const quoteResp = await fetch(`${RELAY_API}/quote`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(15_000),
  });
  const quote = await quoteResp.json();
  if (quoteResp.status !== 200) {
    throw new Error(`Relay quote failed (${quoteResp.status}): ${JSON.stringify(quote)}`);
  }

  let txHash = "";
  const checkEndpoints: string[] = [];
  for (const step of quote.steps ?? []) {
    if (step.kind !== "transaction") continue;
    for (const item of step.items ?? []) {
      const data = item.data ?? {};
      txHash = await sendOriginTx(provider, wallet, origin, data);
      log.info(`[Relay] step '${step.id}' sent on ${origin.name}: ${txHash}`);
      if (item.check?.endpoint) checkEndpoints.push(item.check.endpoint);
    }
  }

  if (!txHash) {
    throw new Error(`Relay quote returned no executable transaction steps: ${JSON.stringify(quote)}`);
  }

  for (const endpoint of checkEndpoints) {
    await waitForFill(endpoint);
  }
  return txHash;
}

/** Relay may return numeric fields as int, decimal string, or hex string. */
export function toInt(value: unknown): bigint {
  if (value === null || value === undefined) return 0n;
  if (typeof value === "string") {
    return value.startsWith("0x") ? BigInt(value) : BigInt(parseInt(value, 10));
  }
  return BigInt(value as number);
}

async function sendOriginTx(
  provider: JsonRpcProvider,
  wallet: Wallet,
  origin: EvmChain,
  data: Record<string, any>
): Promise<string> {
  const tx: Record<string, unknown> = {
    to: getAddress(data.to),
    data: data.data ?? "0x",
    value: toInt(data.value),
    chainId: origin.chainId,
  };
  if (origin.eip1559) {
    const block = await provider.getBlock("latest");
    const priority = parseUnits("2", "gwei");
    const baseFee = block?.baseFeePerGas ?? (await provider.getFeeData()).gasPrice ?? 0n;
    tx.maxPriorityFeePerGas = priority;
    tx.maxFeePerGas = baseFee * 2n + priority;
  } else {
    tx.gasPrice = (await provider.getFeeData()).gasPrice;
  }
  const estimated = await provider.estimateGas({ ...tx, from: wallet.address } as any);
  tx.gasLimit = (estimated * 13n) / 10n;

  const sent = await wallet.sendTransaction(tx as any);
  const receipt = await sent.wait(1, 180_000);
  if (!receipt || receipt.status !== 1) {
    throw new Error(`Relay origin transaction reverted: ${sent.hash}`);
  }
  return sent.hash;
}

/** Poll Relay's status endpoint until the cross-chain fill completes. */
async function waitForFill(endpoint: string): Promise<void> {
  const url = endpoint.startsWith("http") ? endpoint : `${RELAY_API}${endpoint}`;
  for (let i = 0; i < STATUS_POLL_TRIES; i++) {
    await sleep(STATUS_POLL_SECONDS * 1000);
    let status = "";
    try {
      const resp = await fetch(url, { signal: AbortSignal.timeout(10_000) });
      status = (await resp.json())?.status ?? "";
    } catch {
      continue;
    }
    if (status === "success") {
      log.info("[Relay] fill confirmed");
      return;
    }
    if (["failure", "refund"].includes(status)) {
      throw new Error(`Relay fill ended with status '${status}'`);
    }
  }
  throw new Error("Relay fill not confirmed in time (check relay.link/transactions)");
}
